docs/content/posts/howto-test-batch-source.md

```diff
@@ -0,0 +1,202 @@
+methods. For
+that we advice the use of [testContainers](https://www.testcontainers.org/) that relies on docker
+images to provide a real backend
+instance (not a mock) that is representative for integration tests. Several backends are supported
+out of the box by testContainers. We need to configure test containers that way:
+
+* Redirect container output (error and standard output) to Flink logs
+* Set the different timeouts to cope with CI server load
+* Set retrial mechanisms for connection, initialization requests etc... for the same reason
+
+### Checkpointing semantics
+
+In big data execution engines, there are 2 levels of guaranty regarding source and sinks:
+
+* At least once: upon failure and recovery, some records may be reflected multiple times but none
+  will
+  be lost
+* Exactly once: upon failure and recovery, every record will be reflected exactly once
+  By the following code we verify that the source supports exactly once semantics:
+
+`@TestSemantics
+CheckpointingMode[] semantics = new CheckpointingMode[] {CheckpointingMode.EXACTLY_ONCE};
+`
+
+That being said, we could encounter a problem while running the tests : the default assertions in
+the Flink source test framework assume that the data is read in the same order it was written. This
+is untrue for most big data backends where ordering is usually not deterministic. To support
+unordered checks and still use all the framework provided tests, we need to override
+[SourceTestSuiteBase#checkResultWithSemantic](https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.html#checkResultWithSemantic-org.apache.flink.util.CloseableIterator-java.util.List-org.apache.flink.streaming.api.CheckpointingMode-java.lang.Integer-)
+in out ITCase:
+
+`@Override
+protected void checkResultWithSemantic(
+CloseableIterator<Pojo> resultIterator,
+List<List<Pojo>> testData,
+CheckpointingMode semantic,
+Integer limit) {
+if (limit != null) {
+Runnable runnable =
+() ->
+CollectIteratorAssertions.assertUnordered(resultIterator)
+.withNumRecordsLimit(limit)
+.matchesRecordsFromSource(testData, semantic);
+
+        assertThat(runAsync(runnable)).succeedsWithin(DEFAULT_COLLECT_DATA_TIMEOUT);
+    } else {
+        CollectIteratorAssertions.assertUnordered(resultIterator)
+                .matchesRecordsFromSource(testData, semantic);
+    }
+
+}`
+
+This is simply a copy-paste of the parent method where _CollectIteratorAssertions.assertOrdered()_
+is
+replaced by _CollectIteratorAssertions.assertUnordered()_.
+
+### Test context
+
+[example Cassandra TestContext](https://github.com/apache/flink-connector-cassandra/blob/d92dc8d891098a9ca6a7de6062b4630079beaaef/flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/CassandraTestContext.java)
+
+The test context is scoped to the test case. So it is where we do things like creating test table,
+creating the source or writing test data.
```

Comment:
```suggestion
The test context provides Flink with means to interact with the backend, like inserting test data/tables or constructing sources/sink.
```

Reply:
Agree with the rephrasing but I'd like also to keep the important notion of test case scope of the TestContext (compared to test suite scope of the backend environment)